### PR TITLE
Implemented nosniff for fetch algorithm

### LINKED ISSUE #14521: Implement X-Content-Type-Options: nosniff

### PR DESCRIPTION
**Pull request by dpyro**
Implemented [nosniff](https://fetch.spec.whatwg.org/#should-response-to-request-be-blocked-due-to-nosniff?) for [fetch algorithm](https://fetch.spec.whatwg.org).

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #14521 (github issue number if applicable).

- [X] There are tests for these changes

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14620)

### CODE REVIEW

File: components/net/fetch/methods.rs
Comment (KiChjang) [inline components/net/fetch/methods.rs]: This is **NOT** an improvement. We purposely avoided cloning because it's an expensive operation to do so for a Response.

File: components/net/fetch/methods.rs
Comment (KiChjang) [inline components/net/fetch/methods.rs]: This function is big enough to be on the top level.

**Comment by dpyro** [comment]
@KiChjang I’ve given it another go using `std::borrow::Cow`. I’ve still removed the inner scope for Steps 14-18 and instead have used only immutable shadows. Media top-level mime types are also blocked for font requests.

File: components/net/fetch/methods.rs
Comment (KiChjang) [inline components/net/fetch/methods.rs]: There's still a clone here.

File: components/net/fetch/methods.rs
Comment (KiChjang) [inline components/net/fetch/methods.rs]: What's the purpose of this?

**Comment by dpyro** [comment]
I rewrote the code to defer mutating *response* so that *internalResponse* could simply borrow it. There may be a `clone()` issue using `into_owned()` for *internalResponse*’s borrow on `response.actual_response()` when assigning it to the `response.internal_response` field. What should be done there?

**Comment by KiChjang** [comment]
@dpyro I don't think going the route of `Cow` is the correct way here; we're just hiding cloning issues behind the struct. Instead, I'd like to know clearly what exact borrow issues that you were having if you simply use references.

**Comment by dpyro** [comment]
@KiChjang I resolved my borrow issues using an inner scope to release a borrow and going through the Rustonomicon to better understand the borrow system. I will be able to use explicit borrows now. 

Do we need to populate `response.internal_response` with *internalResponse* when returning *response* from the main_fetch? If not, avoiding `clone()` should be straightforward.

**Comment by KiChjang** [comment]
I think the problematic part that you're mostly having is when the spec tells you to "set response and internalResponse to a network error". It can be done by doing `response = Response::network_error();` and then `response.internal_response = Response::network_error()`, can it not?

**Comment by dpyro** [comment]
That part is already done with `Response::network_error(error.clone())`. The trouble I’m having is that `internal_response` borrows `response.actual_response()` when there are no errors, so I can’t mutate `response.internal_response` to contain `internal_response` without `clone()` or an implied equivalent.

**Comment by KiChjang** [comment]
I don't seem to see a need to reassign *internalResponse* back to `response.internal_response`, aside from the case where it's supposed to set network errors on both. Step 15 looks like one, but mutating *internalResponse* fields can easily be done by calling `borrow_mut()` or `set()` on the fields themselves.

**Comment by KiChjang** [comment]
In other words, *internalResponse* is there to keep a reference to `response.internal_response`, and is one of the reasons why I highly opposed cloning the actual response.

**Comment by dpyro** [comment]
Just updated the PR. I believe I’ve removed all response cloning by not reassigning *internalResponse* when it’s not a network error.

File: components/net/fetch/methods.rs
Comment (KiChjang) [inline components/net/fetch/methods.rs]: This is a mind twister, and it either requires documentation, or a [bug filed against the spec](https://github.com/whatwg/fetch), or both. Let's say both.

File: components/net/fetch/methods.rs
Comment (KiChjang) [inline components/net/fetch/methods.rs]: nit: Unnecessary type annotation.

File: components/net/fetch/methods.rs
Comment (dpyro) [inline components/net/fetch/methods.rs]: It’s equivalent. However, it’s an artifact from a previous effort and is now reverted.

File: components/net/fetch/methods.rs
Comment (KiChjang) [inline components/net/fetch/methods.rs]: Sadly, I don't think this is equivalent. In Step 16, you were supposed to set response into a network error if `should_block_nosniff` returns true, and since only `internal_response` is mutated, this check would erroneously think response is not a network error.

Comment (dpyro) [inline components/net/fetch/methods.rs]: You are correct. We should add a test for that, since currently it will pass the unit tests either way. I will see if I can add something.

File: components/net/fetch/methods.rs
Comment (KiChjang) [inline components/net/fetch/methods.rs]: This hasn't been addressed.

**Comment by dpyro** [comment]
Any feedback on this, @KiChjang?

File: components/net/fetch/methods.rs
Comment (KiChjang) [inline components/net/fetch/methods.rs]: We really should avoid writing long comments because it would easily confuse future maintainers. Perhaps the easier way would be to check both `response` and `internal_response` and see if any of the two is a network error, and then leave a short comment there saying that `internal_response` needs to be checked because step 16 did not modify `response` to be a network error.

**Comment by KiChjang** [comment]
Changes look good to me after addressing comments. cc @jdm for the final approval.

File: components/net/fetch/methods.rs
Comment (jdm) [inline components/net/fetch/methods.rs]: Let's remove these inline annotations - they only matter for code that is performance-sensitive, and this code is not.

File: components/net/fetch/methods.rs
Comment (jdm) [inline components/net/fetch/methods.rs]: Let's do:
```rust
let header = match response.headers.get::<XContentTypeOptions>() {
    Some(h) => h,
    None => return false,
};
```
to avoid the extra level of indentation.

File: components/net/fetch/methods.rs
Comment (jdm) [inline components/net/fetch/methods.rs]: Same transformation as previous to avoid indentation.

File: components/net/fetch/methods.rs
Comment (jdm) [inline components/net/fetch/methods.rs]: Where are these steps coming from? They don't appear in https://fetch.spec.whatwg.org/#should-response-to-request-be-blocked-due-to-nosniff? at all.

**Review by jdm — CHANGES_REQUESTED** [review]
In general this looks good!

**Comment by dpyro** [comment]
I’ve been off for the holidays! Will clean this up now.

File: components/net/fetch/methods.rs
Comment (dpyro) [inline components/net/fetch/methods.rs]: It was changed after I made the implementation here: https://github.com/whatwg/fetch/commit/169de91ca9fa3ab91a860bc492caf5fa94c29592

**Comment by dpyro** [comment]
Rebased and added @jdm’s suggestions.

**Comment by jdm** [comment]
@bors-servo: r+
Great work!

**Comment by jdm** [comment]
```

  ▶ Unexpected subtest result in /fetch/nosniff/script.html:
  └ PASS [expected FAIL] URL query: ?type=x/x

  ▶ Unexpected subtest result in /fetch/nosniff/worker.html:
  └ PASS [expected FAIL] URL query: ?type=x/x

  ▶ Unexpected subtest result in /fetch/nosniff/stylesheet.html:
  └ PASS [expected FAIL] URL query: ?type=x/x
```
Excellent, I like tests that start passing with this change. @dpyro you'll need to update script.html.ini, worker.html.ini, and stylesheet.html.ini to remove those specific passing subtests. It would be worth figuring out why we don't pass the remaining subtests in those files as well.

**Comment by dpyro** [comment]
All three of those subtests use `fails = ["", "?type=", "?type=x", "?type=x/x”]`. Only the last one is a valid MIME type. 

In the [nosniff spec](https://fetch.spec.whatwg.org/#should-response-to-request-be-blocked-due-to-nosniff%3F), Step 4 assumes that we can extract a valid MIME type from _response_’s header list.

I’m not sure what the valid behavior should be for this case, so I simply returned an allow if we are unable to match on `response.headers.get::<ContentType>()`.

**Comment by Ms2ger** [comment]
It looks like ["extract a MIME type"] returns the empty byte sequence in that case. I think that means you need to block for script and style, and allow in all other cases. It might be easier to start the function with

```rust
let is_script = match type_ {
    Type::Script => true,
    Type::Style => false,
    _ => return false,
};
```

to avoid doing all the unnecessary work in the non-script, non-style case.

["extract a MIME type"]: https://fetch.spec.whatwg.org/#concept-header-extract-mime-type

**Comment by dpyro** [comment]
@jdm @Ms2ger: Currently investigating which tests should pass and which should fail! I’m digging through the `hyper` crate to verify the logic. I have some questions:

- Where do the web platform tests come from? Are pass and fails supposed to be the ultimate correct result, or just the current state of servo?
- Is `Content-Type`/`X-Content-Type-Options` equivalent to `X-Content-Type`/`Content-Type-Options`?
- Is the value of `“nosniff”` valid? Is `nosniff` valid?
- Is `x` am invalid MIME type? (I believe it is not, but I’d like to confirm).

I’ll be investigating these on my own regardless, but I’d appreciate any insight to help speed my progress!

**Comment by jdm** [comment]
Passes and fails reflect the current state of Servo. Ultimately, if the test is correct, all tests should ideally be passing when the features they test are implemented correctly in Servo. Web platform tests come from https://github.com/w3c/web-platform-tests/ which is shared with other browser vendors.

**Comment by jdm** [comment]
As to your second question, Content-Type is an officially specified header, and X-Content-Type-Options is a quasi-standardized header. X-Content-Type is not meaningful, and there is no standardized Content-Type-Options header. The X- prefix is used to indicate nonstandard, application specific headers in general.

**Comment by dpyro** [comment]
Thanks for the info, @jdm. Finally understood the web-platform-tests and figured out how to wrangle `mach`! It should now only FAIL on URL query but I am investigating it.

**Comment by dpyro** [comment]
@jdm: I keep receiving intermittent failures when running `./mach test-wpt tests/wpt/web-platform-tests/fetch/` on my branch. What changes could I have made that could have caused this?

```
$ ./mach test-wpt tests/wpt/web-platform-tests/fetch/nosniff/
Running 6 tests in web-platform-tests

  ▶ Unexpected subtest result in /fetch/nosniff/parsing-nosniff.html:
  │ FAIL [expected PASS] URL query: uppercase
  └   → assert_unreached: Unexpected load event Reached unreachable code

  ▶ Unexpected subtest result in /fetch/nosniff/parsing-nosniff.html:
  │ FAIL [expected PASS] URL query: last
  └   → assert_unreached: Unexpected error event Reached unreachable code

  ▶ Unexpected subtest result in /fetch/nosniff/parsing-nosniff.html:
  │ FAIL [expected PASS] URL query: quoted
  └   → assert_unreached: Unexpected error event Reached unreachable code

Ran 6 tests finished in 5.0 seconds.
  • 5 ran as expected. 0 tests skipped.
  • 1 tests had unexpected subtest results

$ ./mach test-wpt tests/wpt/web-platform-tests/fetch/nosniff/
Running 6 tests in web-platform-tests

Ran 6 tests finished in 4.0 seconds.
  • 6 ran as expected. 0 tests skipped.

$ ./mach test-wpt tests/wpt/web-platform-tests/fetch/nosniff/
Running 6 tests in web-platform-tests

Ran 6 tests finished in 4.0 seconds.
  • 6 ran as expected. 0 tests skipped.

$ ./mach test-wpt tests/wpt/web-platform-tests/fetch/nosniff/
Running 6 tests in web-platform-tests

Ran 6 tests finished in 4.0 seconds.
  • 6 ran as expected. 0 tests skipped.

$ ./mach test-wpt tests/wpt/web-platform-tests/fetch/nosniff/
Running 6 tests in web-platform-tests

  ▶ Unexpected subtest result in /fetch/nosniff/parsing-nosniff.html:
  │ FAIL [expected PASS] URL query: uppercase
  └   → assert_unreached: Unexpected load event Reached unreachable code

  ▶ Unexpected subtest result in /fetch/nosniff/parsing-nosniff.html:
  │ FAIL [expected PASS] URL query: last
  └   → assert_unreached: Unexpected error event Reached unreachable code

  ▶ Unexpected subtest result in /fetch/nosniff/parsing-nosniff.html:
  │ FAIL [expected PASS] URL query: quoted
  └   → assert_unreached: Unexpected error event Reached unreachable code

Ran 6 tests finished in 4.0 seconds.
  • 5 ran as expected. 0 tests skipped.
  • 1 tests had unexpected subtest results
```

**Comment by dpyro** [comment]
I believe I’ve gotten it all to work now! I did have to `impl` a custom `hyper::header::Header` to parse only the first header according to the spec.

File: components/net/fetch/methods.rs
Comment (jdm) [inline components/net/fetch/methods.rs]: nit: || should be on the previous line.

**Review by jdm — CHANGES_REQUESTED** [review]
Only one tiny little stylistic nit. This is great work; thanks for sticking with it!

**Comment by jdm** [comment]
```
  ▶ Unexpected subtest result in /fetch/nosniff/image.html:
  │ FAIL [expected PASS] URL query: 
  │   → assert_equals: expected 96 but got 16
  │ FAIL [expected PASS] URL query: ?type=
  │   → assert_equals: expected 96 but got 16
  │ FAIL [expected PASS] URL query: ?type=x
  │   → assert_equals: expected 96 but got 16
  │ 
  │ img.onload<@http://web-platform.test:8000/fetch/nosniff/image.html:13:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1446:17

  ▶ Unexpected subtest result in /fetch/nosniff/parsing-nosniff.html:
  │ FAIL [expected PASS] URL query: uppercase
  └   → assert_unreached: Unexpected load event Reached unreachable code
```
Curious.

**Comment by jdm** [comment]
parsing-nosniff.html continues to have odd results:
```
  ▶ Unexpected subtest result in /fetch/nosniff/parsing-nosniff.html:
  │ FAIL [expected PASS] URL query: uppercase
  └   → assert_unreached: Unexpected load event Reached unreachable code
```

**Comment by jdm** [comment]
That appears to happen consistently on both mac and linux.

**Comment by jdm** [comment]
Whereas running the test locally, I consistently see these failures instead:
```
/fetch/nosniff/parsing-nosniff.html
-----------------------------------
FAIL URL query: quoted
FAIL URL query: no-x
```

**Comment by jdm** [comment]
I found the issue, and it's not caused by this code. #15694 is the problem here.

**Comment by jdm** [comment]
We'll hold this one back until #15696 merges.

**Comment by dpyro** [comment]
So that’s what was going on! The failures were intermittent but only happened on my branch, which left me very confused!

**Comment by jdm** [comment]
```
 ▶ Unexpected subtest result in /workers/WorkerGlobalScope_importScripts_NosniffErr.htm:
  └ PASS [expected FAIL] importScripts throws on 'nosniff' violation

  ▶ OK [expected TIMEOUT] /workers/Worker_NosniffErr.htm

  ▶ Unexpected subtest result in /workers/Worker_NosniffErr.htm:
  └ PASS [expected TIMEOUT]  Worker with nosniff X-Content-Type-Options header 
```
Looks like two final text expectations to update!

**Comment by nox** [comment]
@jdm Only expectation changes remain indeed, stealing it nonetheless to free you a bit more.

**Pull request merged**
